File: leetcode3/이진희/315. Count of Smaller Numbers After Self.java
/*

1. 아이디어 : 머지 소트 방식 활용. 병합정렬중 Merge 단계에서, 왼쪽 배열의 값이 오른쪽 배열의 값보다 크다면, 
             오른쪽 배열의 값이 왼쪽 배열의 값보다 작다는 의미이므로, count를 증가시킨다

2. 시간복잡도 : O(NlogN)

3. 자료구조/알고리즘 : 머지소트

 */

class Solution {
    private int[] count;

    private class Pair {
        int val;
        int index;

        Pair(int val, int index) {
            this.val = val;
            this.index = index;
        }
    }

    public List<Integer> countSmaller(int[] nums) {
        int n = nums.length;
        count = new int[n];
        
        Pair[] pairs = new Pair[n];
        for (int i = 0; i < n; i++) {
            pairs[i] = new Pair(nums[i], i);
        }

        mergeSort(pairs, 0, n - 1);

        List<Integer> result = new ArrayList<>();
        for (int c : count) {
            result.add(c);
        }
        return result;
    }

    private void mergeSort(Pair[] pairs, int left, int right) {
        if (left >= right) return;

        int mid = left + (right - left) / 2;
        mergeSort(pairs, left, mid);
        mergeSort(pairs, mid + 1, right);

        merge(pairs, left, mid, right);
    }

    private void merge(Pair[] pairs, int left, int mid, int right) {
        Pair[] temp = new Pair[right - left + 1];
        int i = left;
        int j = mid + 1;
        int k = 0;
        int rightCount = 0;

        while (i <= mid && j <= right) {
            if (pairs[j].val < pairs[i].val) {
                rightCount++;
                temp[k++] = pairs[j++];
            } else {
                count[pairs[i].index] += rightCount;
                temp[k++] = pairs[i++];
            }
        }

        while (i <= mid) {
            count[pairs[i].index] += rightCount;
            temp[k++] = pairs[i++];
        }

        while (j <= right) {
            temp[k++] = pairs[j++];
        }

        System.arraycopy(temp, 0, pairs, left, temp.length);
    }
}

File: leetcode3/김민제/315. Count of Smaller Numbers After Self.py
from typing import List


class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:

        n = len(nums)
        counts = [0] * n

        # (값, 원래 index)
        arr = [(nums[i], i) for i in range(n)]

        def merge_sort(left, right):

            if left >= right:
                return

            mid = (left + right) // 2

            merge_sort(left, mid)
            merge_sort(mid + 1, right)

            temp = []

            i = left
            j = mid + 1

            # 오른쪽에서 이미 가져온 작은 원소 개수
            right_smaller = 0

            while i <= mid and j <= right:

                if arr[j][0] < arr[i][0]:
                    temp.append(arr[j])
                    right_smaller += 1
                    j += 1

                else:
                    # arr[i]보다 작은 오른쪽 원소들의 개수
                    counts[arr[i][1]] += right_smaller

                    temp.append(arr[i])
                    i += 1

            # 왼쪽에 남은 원소
            while i <= mid:
                counts[arr[i][1]] += right_smaller

                temp.append(arr[i])
                i += 1

            # 오른쪽에 남은 원소
            while j <= right:
                temp.append(arr[j])
                j += 1

            # 정렬된 결과를 원래 배열에 반영
            arr[left:right + 1] = temp

        merge_sort(0, n - 1)

        return counts
